### PR TITLE
Add HTTP specification

### PR DESCRIPTION
**Pull request by chamil321**
## Purpose
Related to https://github.com/ballerina-platform/ballerina-standard-library/issues/572

## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests

### CODE REVIEW

File: docs/spec/spec.md [hunk 1]
```diff
@@ -0,0 +1,1555 @@
...
+constructor, the listener opens the port and attaches the upgrader service at the given service path.
+When the http:Listener is passed in, the upgrader service gets attached to it at the given service path. The service 
+by default attaches to “/”.
+
+
+```ballerina
+http:Listener hl = new(9090);
+
+service /ws on new ws:Listener(hl) {
+    remote function onUpgrade(http:Caller hc, http:Request req) returns ws:Service|ws:UpgradeError {
```
Comment (Bhashinee) [inline docs/spec/spec.md:1544]:
```suggestion
    resource function get .(http:Request req) returns websocket:Service|websocket:UpgradeError {
```

File: docs/spec/spec.md [hunk 2]
```diff
@@ -0,0 +1,1555 @@
...
+        return new MyWSService();
+    }
+}
+
+service class MyWSService {
+    *ws:Service;
+    remote function onText(ws:Caller caller, string text) {
```
Comment (Bhashinee) [inline docs/spec/spec.md:1551]:
```suggestion
    *websocket:Service;
    remote function onTextMessage(websocket:Caller caller, string data) returns websocket:Error? {
```

File: docs/spec/spec.md [hunk 3]
```diff
@@ -0,0 +1,1555 @@
...
+        caller->pushText(text);
```
Comment (Bhashinee) [inline docs/spec/spec.md:1552]:
```suggestion
    *websocket:Service;
    remote function onTextMessage(websocket:Caller caller, string data) returns websocket:Error? {
        check caller->writeTextMessage(data);
```

File: docs/spec/spec.md [hunk 4]
```diff
@@ -1459,10 +1468,166 @@ parseHeader(string headerValue) returns HeaderValue[]|ClientError  {
...
+        }
+    }
+
+    public isolated function remove(string 'key) {
+        lock {
+            value:Cloneable|isolated object {} _ = self.attributes.remove('key);
+        }
+    }
+
+    public isolated function next() returns error? = external;
```
Comment (TharmiganK) [inline docs/spec/spec.md:1550]: `FilterContext`->`next()` should return `NextService|error?` right?

File: docs/spec/spec.md [hunk 5]
```diff
@@ -1459,10 +1468,166 @@ parseHeader(string headerValue) returns HeaderValue[]|ClientError  {
...
+the execution flow. Users must invoke next() method in order to trigger the next interceptor in the pipeline. 
+Previously, this was controlled by returning a boolean value which is quite cryptic and confusing.
+
+Moreover, invoking next() method results in immediately executing the next interceptor in the pipeline. Which means 
+whatever below the next() method will only get executed after the completion of execution of the next Interceptor.
+
+If the users forget to call next() and don't respond either, the request will be left hanging.
+
+##### 8.1.1.3 Returning error?
+Resource functions can return only error?. In the case of an error, interceptor chain execution jumps to the last
```
Comment (TharmiganK) [inline docs/spec/spec.md:1565]: Resource functions can return `NextService|error?`. They can return only `error?` **when using the `Caller`**. 
Also, in case of an error, interceptor pipeline execution jumps to the **nearest `RequestErrorInterceptor`** in the pipeline. This `RequestErrorInterceptor` is not necessarily the last interceptor in the pipeline, it can be anywhere in the chain.

File: docs/spec/spec.md [hunk 6]
```diff
@@ -1459,10 +1468,166 @@ parseHeader(string headerValue) returns HeaderValue[]|ClientError  {
...
+ 
+   resource function 'default foo(http:FilterContext ctx, http:Caller caller,
+                       http:Request req) returns error? {
+       request.setHeader("X-requestHeader", "RequestInterceptor");
+       ctx.next();
+   }
+}
+```
+
+##### 8.1.1.1 FilterContext
```
Comment (TharmiganK) [inline docs/spec/spec.md:1519]: Don't we use the `RequestContext` term rather than `FilterContext`?

File: docs/spec/spec.md [hunk 7]
```diff
@@ -1459,10 +1468,166 @@ parseHeader(string headerValue) returns HeaderValue[]|ClientError  {
...
+       // deal with the error
+   }
+}
+```
+
+#### 8.1.3 Engaging Interceptors
+##### 8.1.3.1 Service Level
+Interceptors could get engaged at service level. One reason for this is that users may 
+want to engage two different interceptor chains for each service even though it is attached to the same Listener. At the 
+listener level resource function paths are relative to the service base path.
```
Comment (TharmiganK) [inline docs/spec/spec.md:1594]: At the **service** level

File: docs/spec/spec.md [hunk 8]
```diff
@@ -1459,10 +1468,166 @@ parseHeader(string headerValue) returns HeaderValue[]|ClientError  {
...
+
+##### 8.1.3.2 Listener Level
+Interceptors could get engaged at Listener level as well. At the listener level resource function paths are 
+relative to the /.
+```ballerina
+listener http:Listener echoListener = new http:Listener(9090, config = {interceptors: [requestInterceptor, responseInterceptor]});
+```
+
+##### 8.1.3.3 Execution Order of Interceptors
+These interceptors are always there in the pipeline and executed as the last.
```
Comment (TharmiganK) [inline docs/spec/spec.md:1609]: This line is not related to the topic

File: docs/spec/spec.md [hunk 9]
```diff
@@ -1459,10 +1468,166 @@ parseHeader(string headerValue) returns HeaderValue[]|ClientError  {
...
+
+![img.png](img.png)
+In the above example blue dashed box represents the RequestErrorInterceptor and blue boxes simply represent the 
+RequestInterceptors whereas green dashed box represents the ResponseErrorInterceptor(TBA) and green boxes simply represent the 
+ResponseInterceptors. The new execution orders is as follows,
+```
+RequestInterceptor/RequestErrorInterceptor: 1, 2, 4, 5
+ResponseInterceptor/ResponseErrorInterceptor: 5, 3, 0
+```
+
```
Comment (TharmiganK) [inline docs/spec/spec.md:1619]: Do we need to say about the execution order when interceptors are engaged at both service and listener levels?

`ListenerLevel : RequestInterceptors` -> `ServiceLevel : RequestInterceptors` -> `TargetService` -> `ServiceLevel : ResponseInterceptors` -> `ListenerLevel : ResponseInterceptors`

File: docs/spec/spec.md [hunk 10]
```diff
@@ -0,0 +1,1821 @@
...
+    * 10.2. [Websocket](#102-websocket)
+
+## 1. Overview
+This specification elaborates on Basic Auth authentication and authorization for all the Ballerina listeners and
+clients. The HTTP, gRPC, GraphQL, WebSocket, WebSub protocol-based listeners and clients are secured according to this
+specification.
+
+The ballerina language provides the first class support for writing network oriented programming models in a more 
+initiative manner. The ballerina-http standard library consumes the rich language construct and creates the 
+programming model to write services.
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:123]: I feel the start can be improved. How about something like this ?

The HTTP standard library is designed to work with HTTP protocol. It includes high level abstractions such as `http:Request`, `http:Response`, `http:Service` and `http:Client` which allow users to consume and produce HTTP services. Further the users can use this library to build other libraries. In fact, the standard libraries such as GraphQL, Websub and WebSubHub use this library internally.  

In addition to functional requirements, this library deals with none functional requirements such as security, observerbility and resiliency. Each requirement is discussed in detail in the coming sections.

File: docs/spec/spec.md [hunk 11]
```diff
@@ -0,0 +1,1821 @@
...
+## 2. Components
+### 2.1. Listener
+The HTTP listener object receives network data from a remote process according to the HTTP transport protocol and 
+translates the received data into invocations on the resources functions of services that have been 
+attached to the listener object. The listener provides the interface between network and services.
+
+As defined in [Ballerina 2021R1 Section 5.7.4](https://ballerina.io/spec/lang/2021R1/#section_5.7.4) the Listener has 
+the object constructor and life cycle methods such as attach(), detach(), 'start(), gracefulStop(), and immediateStop().
+
+#### 2.1.1. Auto binding
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:134]: I feel `binding` is not the correct word here.

Comment (shafreenAnfar) [reply docs/spec/spec.md:134]: 2.1.1. Automatically starting the service
2.1.2. Programmatically starting the service

Comment (shafreenAnfar) [reply docs/spec/spec.md:134]: Above two might be a better option.

File: docs/spec/spec.md [hunk 12]
```diff
@@ -0,0 +1,1821 @@
...
+}
+```
+
+#### 2.1.2. Late binding
+
+Users can programmatically start the listener by calling each lifecycle method as follows.
+
+```ballerina
+// Listener object constructor
+listener http:Listener serviceListener = new(9090);
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:156]: Let's create the listener inside the main as well.

File: docs/spec/spec.md [hunk 13]
```diff
@@ -0,0 +1,1821 @@
...
+[listener declaration](https://ballerina.io/spec/lang/2021R1/#section_8.3.1)
+
+```ballerina
+// Listener object constructor
+listener http:Listener serviceListener = new(9090, config = {httpVersion: "2.0"});
+```
+```ballerina
+// Service attaches to the Listener
+service http:Service /sessionTest on serviceListener {
+
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:146]: Maybe we can add the same.

`resource function get sayHello(http:Caller caller) {}`

File: docs/spec/spec.md [hunk 14]
```diff
@@ -0,0 +1,1821 @@
...
+In addition to that a service can contain public and private functions which can be accessed by calling with `self`.
+
+#### 2.2.1. Service type
+```ballerina
+public type Service distinct service object {
+
+};
+```
+Above distinct type is provided by HTTP module and user can include the type as `*http:Service` to refer it.
+The comprehensive typing support is yet to be added to the language
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:181]:
```suggestion
The comprehensive typing support is yet to be added to the language.
```

Comment (shafreenAnfar) [reply docs/spec/spec.md:181]: I think better to add a note say until that compiler plugins are used to validate the services.

File: docs/spec/spec.md [hunk 15]
```diff
@@ -0,0 +1,1821 @@
...
+
+#### 2.2.2. Service base path
+
+The base path is considered during the request dispatching to discover the service. Identifiers and string literals
+are allowed to be stated as base path and it should be started with `/`. The base path is optional and it will be 
+default to `/` when not defined. If the base path contains any special charactors, those should be escaped or defined
+as string literals
+
+```ballerina
+service http:Service /my/Tes\@tHello/go new http:Listener(9090) {
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:191]: we can remove `http:Service` add something like `hello\-world` for the path.

File: docs/spec/spec.md [hunk 16]
```diff
@@ -0,0 +1,1821 @@
...
+```ballerina
+service isolated class SClass {
+   *http:Service;
+   isolated resource function get greeting() returns string {
+       return "hello world";
+   }
+}
+
+listener http:Listener serviceListener = new (9090);
+http:Service httpService = new SClass();
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:233]: We can have the above two inside the main.

File: docs/spec/spec.md [hunk 17]
```diff
@@ -0,0 +1,1821 @@
...
+}
+```
+
+#### 2.2.5. Service constructor expression
+
+```ballerina
+listener http:Listener serviceListener = new (9090);
+
+http:Service httpService = @http:ServiceConfig {} service object {
+   resource function get baz() returns string {
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:248]: better to use the same resource in all of them.

File: docs/spec/spec.md [hunk 18]
```diff
@@ -0,0 +1,1821 @@
...
+Only the identifiers can be used as resource path not string literals. Dot identifier is 
+used to denote the `/` only if the path contains a single identifier. 
+```ballerina
+resource function post .() {
+    
+}
+```
+Any special characters can be used in the path by escaping.
+```ballerina
+resource function post test\@hello() {
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:296]: `hello\-world` would be better.

File: docs/spec/spec.md [hunk 19]
```diff
@@ -0,0 +1,1821 @@
...
+```ballerina
+resource function post foo(@http:CallerInfo {respondType:Person}  http:Caller hc) {
+    Person p = {};
+    hc->respond(p);
+}
+```
+##### 2.3.4.2. http:Request
+
+The `http:Request` represents the message which came along over the network that modelled as an object along with 
+headers and payload. Listener passes it to the resource function as an argument to be accessed by the user based on
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:400]: I think the first sentence can be changed to something like the below.

The `http:Request` represents the request which is sent and received over the network which includes headers and the entity body.

File: docs/spec/spec.md [hunk 20]
```diff
@@ -0,0 +1,2305 @@
...
+        }
+    ]
+}
+service /foo on new http:Listener(9090) {
+    resource function get bar() returns string {
+        return "Hello, World!";
+    }
+}
+```
+```ballerina
```
Comment (ldclakmal) [inline docs/spec/spec.md:1741]:
```suggestion

```ballerina
```

File: docs/spec/spec.md [hunk 21]
```diff
@@ -0,0 +1,2305 @@
...
+}
+service /foo on new http:Listener(9090) {
+    resource function get bar() returns string {
+        return "Hello, World!";
+    }
+}
+```
+
+##### 9.1.1.4 Listener - OAuth2
+```ballerina
```
Comment (ldclakmal) [inline docs/spec/spec.md:1823]:
```suggestion

```ballerina
```

File: docs/spec/spec.md [hunk 22]
```diff
@@ -0,0 +1,2305 @@
...
+}
+service /foo on new http:Listener(9090) {
+    resource function get bar() returns string {
+        return "Hello, World!";
+    }
+}
+```
+
+##### 9.1.1.5 Client - Basic Auth
+```ballerina
```
Comment (ldclakmal) [inline docs/spec/spec.md:1850]:
```suggestion

```ballerina
```

File: docs/spec/spec.md [hunk 23]
```diff
@@ -0,0 +1,2305 @@
...
+);
+
+public function main() returns error? {
+    json response = check c->get("/foo/bar");
+    // evaluate response
+}
+```
+
+##### 9.1.1.6 Client - Bearer Token Auth
+```ballerina
```
Comment (ldclakmal) [inline docs/spec/spec.md:1865]:
```suggestion

```ballerina
```

File: docs/spec/spec.md [hunk 24]
```diff
@@ -0,0 +1,2305 @@
...
+);
+
+public function main() returns error? {
+    json response = check c->get("/foo/bar");
+    // evaluate response
+}
+```
+
+##### 9.1.1.7 Client - Self-Signed JWT
+```ballerina
```
Comment (ldclakmal) [inline docs/spec/spec.md:1881]:
```suggestion

```ballerina
```

File: docs/spec/spec.md [hunk 25]
```diff
@@ -0,0 +1,2305 @@
...
+    http:Request req = new;
+    req = check handler->enrich(req);
+    json response = check c->get("/foo/bar");
+    // evaluate response
+}
+```
+
+### 9.2 SSL/TLS and Mutual SSL
+
+The HTTPS listener could connect to or interact with an HTTPS client. The http:ListenerSecureSocket configuration
```
Comment (ldclakmal) [inline docs/spec/spec.md:2163]:
```suggestion
The HTTPS listener could connect to or interact with an HTTPS client. The `http:ListenerSecureSocket` configuration
```

File: docs/spec/spec.md [hunk 26]
```diff
@@ -0,0 +1,2305 @@
...
+of the listener exposes the HTTPS connection related configs.
+
+#### 9.2.1 Listener - SSL/TLS
+```ballerina
+listener http:Listener securedEP = new(9090,
+    secureSocket = {
+        key: {
+            certFile: "../resource/path/to/public.crt",
+            keyFile: "../resource/path/to/private.key"
```
Comment (ldclakmal) [inline docs/spec/spec.md:2172]:
```suggestion
            certFile: "/path/to/public.crt",
            keyFile: "/path/to/private.key"
```

File: docs/spec/spec.md [hunk 27]
```diff
@@ -0,0 +1,2305 @@
...
+The mutual SSL support which is a certificate-based authentication process in which two parties 
+(the client and server) authenticate each other by verifying the digital certificates. It ensures that both 
+parties are assured of each other’s identity.
+
+```ballerina
+listener http:Listener securedEP = new(9090,
+    secureSocket = {
+        key: {
+            certFile: "../resource/path/to/public.crt",
+            keyFile: "../resource/path/to/private.key"
```
Comment (ldclakmal) [inline docs/spec/spec.md:2195]:
```suggestion
            certFile: "/path/to/public.crt",
            keyFile: "/path/to/private.key"
```

File: docs/spec/spec.md [hunk 28]
```diff
@@ -0,0 +1,2305 @@
...
+        },
+
+        mutualSsl: {
+            verifyClient: http:REQUIRE,
+            cert: "../resource/path/to/public.crt"
```
Comment (ldclakmal) [inline docs/spec/spec.md:2200]:
```suggestion
            cert: "/path/to/public.crt"
```

File: docs/spec/spec.md [hunk 29]
```diff
@@ -0,0 +1,2305 @@
...
+    }
+}
+```
+
+#### 9.2.3 Client - SSL/TLS
+
+```ballerina
+http:Client securedEP = check new("https://localhost:9090",
+    secureSocket = {
+        cert: "../resource/path/to/public.crt"
```
Comment (ldclakmal) [inline docs/spec/spec.md:2224]:
```suggestion
        cert: "/path/to/public.crt"
```

File: docs/spec/spec.md [hunk 30]
```diff
@@ -0,0 +1,2305 @@
...
+```
+
+#### 9.2.4 Client - Mutual SSL
+
+```ballerina
+http:Client securedEP = check new("https://localhost:9090",
+    secureSocket = {
+        key: {
+            certFile: "../resource/path/to/public.crt",
+            keyFile: "../resource/path/to/private.key"
```
Comment (ldclakmal) [inline docs/spec/spec.md:2241]:
```suggestion
            certFile: "/path/to/public.crt",
            keyFile: "/path/to/private.key"
```

File: docs/spec/spec.md [hunk 31]
```diff
@@ -0,0 +1,2305 @@
...
+        },
+        cert: "../resource/path/to/public.crt",
```
Comment (ldclakmal) [inline docs/spec/spec.md:2243]:
```suggestion
        cert: "/path/to/public.crt",
```

File: docs/spec/spec.md [hunk 32]
```diff
@@ -0,0 +1,2305 @@
...
+        * 9.1.1. [Declarative Approach](#911-declarative-approach)
+            * 9.1.1.1. [Listener - File User Store](#9111-listener---file-user-store)
+            * 9.1.1.2. [Listener - LDAP User Store](#9112-listener---ldap-user-store)
+            * 9.1.1.3. [Listener - JWT](#9113-listener---jwt)
+            * 9.1.1.4. [Listener - OAuth2](#9114-listener---oauth2)
+            * 9.1.1.5. [Client - Basic Auth](#9115-client---basic-auth)
+            * 9.1.1.6. [Client - Bearer Token Auth](#9116-client---bearer-token-auth)
+            * 9.1.1.7. [Client - Self Signed JWT Auth](#9117-client---self-signed-jwt)
+            * 9.1.1.8. [Client - Bearer Token OAuth2](#9118-client---bearer-token-oauth2)
+            * 9.1.1.9. [Client - Grant types OAuth2](#9119-client---grant-types-oauth2)
```
Comment (ldclakmal) [inline docs/spec/spec.md:102]:
```suggestion
            * 9.1.1.9. [Client - Grant Types OAuth2](#9119-client---grant-types-oauth2)
```

File: docs/spec/spec.md [hunk 33]
```diff
@@ -0,0 +1,2305 @@
...
+        * 9.1.2 [Imperative Approach](#912-imperative-approach)
+            * 9.1.2.1. [Listener - File User Store](#9121-listener---file-user-store)
+            * 9.1.2.2. [Listener - LDAP User Store](#9122-listener---ldap-user-store)
+            * 9.1.2.3. [Listener - JWT](#9123-listener---jwt)
+            * 9.1.2.4. [Listener - OAuth2](#9124-listener---oauth2)
+            * 9.1.2.5. [Client - Basic Auth](#9125-client---basic-auth)
+            * 9.1.2.6. [Client - Bearer Token Auth](#9126-client---bearer-token-auth)
+            * 9.1.2.7. [Client - Self Signed JWT Auth](#9127-self-signed-jwt)
+            * 9.1.2.8. [Client - Bearer token OAuth2](#9128-client---bearer-token-oauth2)
+            * 9.1.2.9. [Client - Grant types OAuth2](#9129-client---grant-types-oauth2)
```
Comment (ldclakmal) [inline docs/spec/spec.md:112]:
```suggestion
            * 9.1.2.8. [Client - Bearer Token OAuth2](#9128-client---bearer-token-oauth2)
            * 9.1.2.9. [Client - Grant Types OAuth2](#9129-client---grant-types-oauth2)
```

File: docs/spec/spec.md [hunk 34]
```diff
@@ -0,0 +1,2314 @@
...
+```ballerina
+service isolated class SClass {
+   *http:Service;
+   resource function get greeting() returns string {
+       return "hello world";
+   }
+}
+
+public function main() {
+   listener http:Listener serviceListener = new (9090);
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:249]: Can we use `listener` keyword inside main ?

File: docs/spec/spec.md [hunk 35]
```diff
@@ -0,0 +1,2314 @@
...
+    
+}
+```
+
+If the query parameter is not defined in the function signature, then the query param binding does not happen. If a 
+query param of the request URL has no corresponding parameter in the resource function, then that param is ignored. 
+If the parameter is defined in the function, but there is no such query param in the URL, that request will lead 
+to a 400 BAD REQUEST error response unless the type is nilable (string?)
+
+The query param consists of query name and values. Sometimes user may sent query without value(`foo:`). In such
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:454]:
```suggestion
The query param consists of query name and values. Sometimes user may send query without value(`foo:`). In such
```

File: docs/spec/spec.md [hunk 36]
```diff
@@ -0,0 +1,2314 @@
...
+< 
+* Connection #0 to host localhost left intact
+* Closing connection 0
+```
+
+### 2.4. Client
+A client allows the program to send network messages to a remote process according to the HTTP protocol. The fixed 
+remote functions of the client object correspond to distinct network operations defined by the HTTP protocol.
+
+The client init function requires port and configuration to initialize the client.
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:873]:
```suggestion
The client init function requires a valid URL and optional configuration to initialize the client. 
```

File: docs/spec/spec.md [hunk 37]
```diff
@@ -0,0 +1,2314 @@
...
+    CookieConfig? cookieConfig = ();
+    ResponseLimitConfigs responseLimits = {};
+    ClientSecureSocket? secureSocket = ();
+|};
+```
+
+Based on the config, the client object will be accompanied by following client behaviours. Following clients cannot be
+instantiated calling `new`, instead user have to enable the config in the `ClientConfiguration`.
+
+##### 2.4.1.1 Secure client
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:904]: I wonder if the topics should be as follows,

2.4.1.1 Security
2.4.1.2 Caching
2.4.1.3 Redirect

File: docs/spec/spec.md [hunk 38]
```diff
@@ -0,0 +1,2314 @@
...
+        cert: {
+            path: TRUSTSTORE_PATH,
+            password: "ballerina"
+        }
+    }
+);
+```
+
+##### 2.4.1.2 Caching client
+An HTTP caching client implementation which wraps with an HTTP caching layer once `cache` config is enabled.
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:923]: We don't need to talk about implementation details in the spec.

File: docs/spec/spec.md [hunk 39]
```diff
@@ -0,0 +1,2314 @@
...
+
+```ballerina
+
+# Submits an HTTP request to a service with the specified HTTP verb.
+# The submit() function does not give out a http:Response as the result.
+# Rather it returns an http:HttpFuture which can be used to do further interactions with the endpoint.
+remote isolated function submit(string httpVerb, string path, RequestMessage message)
+    returns HttpFuture|ClientError;
+
+# This just pass the request to actual network call.
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:1154]: Instead of `This just pass the request to actual network call.` maybe we can say `Passes the request to actual network call.`

File: docs/spec/spec.md [hunk 40]
```diff
@@ -0,0 +1,2314 @@
...
+remote isolated function  getResponse(HttpFuture httpFuture) returns Response|ClientError;
+
+# This just pass the request to actual network call.
+remote isolated function  hasPromise(HttpFuture httpFuture) returns boolean;
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:1158]: Seems like there are additional spaces.

File: docs/spec/spec.md [hunk 41]
```diff
@@ -0,0 +1,2314 @@
...
+resource function post hello(@http:Payload {mediaType:["application/json", "application/ld+json"]} json payload)  {
+    
+}
+```
+
+During the runtime, the request content-type header is matched against the mediaType field value to validate. If the 
+validation fails, the listener returns an error response with the status code of 415 Unsupported Media Type. 
+Otherwise the dispatching moves forward. 
+
+#### 4.3.2. anydata return value info
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:1328]:
```suggestion
#### 4.3.2. Anydata return value info
```

File: docs/spec/spec.md [hunk 42]
```diff
@@ -0,0 +1,2314 @@
...
+
+
+There are few API level additions when it comes to the HTTP/2 design such as Push promise and promise response.
+#### 10.1.1. Push Promise and Promise Response
+Push Promise and Promise response are the only application level new semantics which are introduced by HTTP2.
+
+Other protocol changes such as streams, messages, frames, request prioritization, flow control, header compression, 
+etc. are all lower level changes that can be handled by the HTTP listener seamlessly from the user.
+
+### 10.2. Websocket
```
Comment (shafreenAnfar) [inline docs/spec/spec.md:2287]: I think this is not needed. As decoupled HTTP from Websocket

**Review by shafreenAnfar — APPROVED** [review]
